### PR TITLE
decoupling reducers recipe

### PR DESCRIPTION
**Pull request by titouancreach**
From: https://github.com/rematch/rematch/issues/282

### CODE REVIEW

**Comment by ShMcK** [comment]
great work! I hope you don't mind if I act like an editor for a bit within this PR.

**Comment by ShMcK** [comment]
What do you think about describing "Decoupling Reducers" as "Reusable Reducers"?

My understanding is "decoupling" refers more to separating two parts, so that neither part knows about each other. The pattern of turning reducers into reusable pure functions doesn't strike me as fitting this definition. What do you think?

File: docs/recipes/decouplingReducers.md [hunk 1]
```diff
@@ -0,0 +1,141 @@
+# Decoupling reducers
+
+Rematch and more generally Redux, encourage to keep a flat state to solve several [issues](https://redux.js.org/recipes/structuring-reducers/normalizing-state-shape).
```
Comment (ShMcK) [inline docs/recipes/decouplingReducers.md]: Rematch, and more generally Redux, encourage you to...

File: docs/recipes/decouplingReducers.md [hunk 2]
```diff
@@ -0,0 +1,141 @@
...
+Each model should be considered as a separate entity and should only be updated by its reducer, this is why Rematch doesn't allow nested reducers.
+
```
Comment (ShMcK) [inline docs/recipes/decouplingReducers.md]: Nested reducers can be used in Rematch, but are not recommended. I think that is an important distinction.

Some users have set up functions, similar to combineReducers, that generate their nested reducers. I've also seen people use redux.reducers with combineReducers.

File: docs/recipes/decouplingReducers.md [hunk 3]
```diff
@@ -0,0 +1,141 @@
...
+This being said, there is some case where nested reducer can be a good thing.
```
Comment (ShMcK) [inline docs/recipes/decouplingReducers.md]: nested reducer**s**

File: docs/recipes/decouplingReducers.md [hunk 4]
```diff
@@ -0,0 +1,141 @@
...
+In this recipe, we will see how to decoupling reducers that tends to grow too much.
+
```
Comment (ShMcK) [inline docs/recipes/decouplingReducers.md]: In this recipe, we will see how to use reusable reducer functions.

File: docs/recipes/decouplingReducers.md [hunk 5]
```diff
@@ -0,0 +1,141 @@
...
+Let's start with an example. A common pattern in Redux, when we want to store an ordered list, is "allIds, byId". Store  the entities as
+an array requires to iterate over the array to find our target but store the entities as an object doesn't let us to keep 
+an order. So let's use both:
```
Comment (ShMcK) [inline docs/recipes/decouplingReducers.md]: "Storing the entities as an array requires an additional cost of iterating over the array to find our target. On the other hand, storing the entities as an object doesn't preserve the order of items"

File: docs/recipes/decouplingReducers.md [hunk 6]
```diff
@@ -0,0 +1,141 @@
...
+      1: {
+        task: 'Learn functional programming',
+        isDone: false
+      }
+    },
+    allIds: [1, 0]
+  }
+}
+```
+`byId` key stores the entity while `allIds` keeps the track of the order of our entities.
```
Comment (ShMcK) [inline docs/recipes/decouplingReducers.md]: "keeps track"

File: docs/recipes/decouplingReducers.md [hunk 7]
```diff
@@ -0,0 +1,141 @@
...
+
+This driven us into an edge case:
```
Comment (ShMcK) [inline docs/recipes/decouplingReducers.md]: This creates another issue:

File: docs/recipes/decouplingReducers.md [hunk 8]
```diff
@@ -0,0 +1,141 @@
...
+   }
+  }
+}
+
+function toggleAllIds(state, payload) {
+  return state;
+}
+```
+
+And we finally mix all together by distibuting a part of the state to our update functions:
```
Comment (ShMcK) [inline docs/recipes/decouplingReducers.md]: "finally mix everything all together"

File: docs/recipes/decouplingReducers.md [hunk 9]
```diff
@@ -0,0 +1,141 @@
...
+      return {
+        byId: toggleById(state.byId, payload),
+        allIds: toogleAllIds(state.allIds, payload)
+      };
+    }
+  }
+};
+
+```
+Our model seems much better and more readable.
```
Comment (ShMcK) [inline docs/recipes/decouplingReducers.md]: "Reusable reducer functions make our model simpler and more readable."

**Comment by titouancreach** [comment]
> Consider these comments largely suggestions to help improve the readability. Overall, the tutorial is quite clear and easy to follow.

Don't worry, I know my writing in English is quite bad and I try to improve it! 

So I take note of what you say to become a better writer :)

**Comment by titouancreach** [comment]
> What do you think about describing "Decoupling Reducers" as "Reusable Reducers" or "Reusable Reducer Functions"?

> My understanding is "decoupling" refers more to separating two parts, so that neither part knows about each other. The pattern of turning reducers into reusable pure functions doesn't strike me as fitting this definition. What do you think?

I think that our case here:
  - allIds doesn't know about byId because we pass state.allIids as an argument
  - byId doesn't know about allIds

File: docs/recipes/decouplingReducers.md (hunk 2 above)
Comment (titouancreach) [reply docs/recipes/decouplingReducers.md]: I would say, this is not possible out of the box: https://github.com/rematch/rematch/issues/282#issuecomment-372644100

**Comment by titouancreach** [comment]
By the way, I didn't test the code I wrote, everything compile in my head, I would probably test it before we merge it

**Comment by ShMcK** [comment]
Definitely give it a test run, if you don't mind. A link to a codesandbox.io demo always helps.

**Comment by titouancreach** [comment]
https://codesandbox.io/s/lry6024mkl This is the example without decoupling (I've to change in the model the key "reducer" to "reducers"

**Comment by titouancreach** [comment]
https://codesandbox.io/s/x2r7nryn24 the second example with decoupling reducers. The mistake I made

Spell "toogle" instead of "toggle"
declare: filterObjectByKey and call filterObjectById

**Comment by titouancreach** [comment]
LGTM, open to feedbacks

**Comment by ShMcK** [comment]
Is it alright if I change the file name to "resuableReducers.md" ?

**Comment by titouancreach** [comment]
If we do that, we need to change the title too. For me, decoupling reducer is good because it's what we are doing

**Comment by titouancreach** [comment]
Thank you :)

**Pull request merged**